lib/ansible/modules/cloud/docker/docker_node.py

```diff
@@ -0,0 +1,338 @@
+          - pause
+          - drain
+    role:
+        description: Node role to assign
+        choices:
+          - manager
+          - worker
+extends_documentation_fragment:
+    - docker
+requirements:
+    - python >= 2.7
+    - "docker >= 2.6.0"
+    - "Please note that the L(docker-py,https://pypi.org/project/docker-py/) Python
+       module has been superseded by L(docker,https://pypi.org/project/docker/)
+       (see L(here,https://github.com/docker/docker-py/issues/1310) for details).
+       Version 2.1.0 or newer is only available with the C(docker) module."
+    - Docker API >= 1.25
+author:
+  - Piotr Wojciechowski (@wojciechowskipiotr)
+  - Thierry Bouvet (@tbouvet)
+
+'''
+
+EXAMPLES = '''
+- name: List names of registered swarm nodes
+  docker_node:
+    state: list
+
+- name: Set node role
+  docker_node:
+    state: update
+    name: mynode
+    role: manager
+
+- name: Set node availability
+  docker_node:
+    state: update
+    name: mynode
+    availability: drain
+
+- name: Set node labels
+  docker_node:
+    state: update
+    name: mynode
+    labels:
+      key: value
+
+- name: Remove node labels
+  docker_node:
+    state: update
+    name: mynode
+    labels_remove: true
+'''
+
+RETURN = '''
+node_facts:
+  description: Information about node after 'update' operation
+  returned: success
+  type: dict
+nodes:
+  description: List nodes names of swarm cluster.
+  returned: success
+  type: list
+  example: "[ 'node-1', 'node-2', 'node-3']"
+actions:
+  description: Provides the actions done on the swarm.
+  returned: when action failed.
+  type: list
+  example: "['This cluster is already a swarm cluster']"
+
+'''
+
+import json
+
+try:
+    from docker.errors import APIError
+except ImportError:
+    # missing docker-py handled in ansible.module_utils.docker_common
+    pass
+
+from ansible.module_utils.docker_common import (
+    AnsibleDockerClient,
+    DockerBaseClass,
+)
+
+from ansible.module_utils._text import to_native
+
+
+class TaskParameters(DockerBaseClass):
+    def __init__(self, client):
+        super(TaskParameters, self).__init__()
+
+        self.state = None
+        self.force = None
+
+        # Spec
+        self.name = None
+        self.labels = None
+        self.labels_remove = None
+
+        # Node
+        self.availability = None
+        self.role = None
+
+        for key, value in client.module.params.items():
+            setattr(self, key, value)
+
+
+class SwarmNodeManager(DockerBaseClass):
+
+    def __init__(self, client, results):
+
+        super(SwarmNodeManager, self).__init__()
+
+        self.client = client
+        self.results = results
+        self.check_mode = self.client.check_mode
+
+        self.parameters = TaskParameters(client)
+
+    def __call__(self):
+        choice_map = {
+            "list": self.node_list,
+            "update": self.node_update,
+        }
+
+        choice_map.get(self.parameters.state)()
+
+    def __isSwarmManager(self):
```

Comment:
That would be a Python module in lib/ansible/module_utils, not a Ansible module (in case this confuses anyone). I guess just naming it `docker_swarm` would be fine, too (everything in module_utils is already expected to be common to multiple modules).

Reply:
But that will require, except changing the name of methods, provide them as argument the `self.client.info()` or `self.client.inspect_swarm()` results or the `Client` object itself to call the method inside. I think the second approach align more with Ansible coding standards

Reply:
From the way the docker modules currently operate, it probably makes sense to add a `AnsibleDockerSwarmClient` to `module_utils/docker_swarm.py`, which extends `AnsibleDockerClient`. Then `docker_swarm`, `docker_swarm_facts`, `docker_node` and `docker_node_facts` could use `AnsibleDockerSwarmClient` instead of using `AnsibleDockerClient` directly.

Reply:
Implemented as new class now